packages/optimizer/lib/transformers/RewriteAmpUrls.js

```diff
@@ -67,25 +67,46 @@ class RewriteAmpUrls {
 
     let node = head.firstChild;
     let referenceNode = findMetaViewport(head);
+    const esm = this.esmModulesEnabled || params.experimentEsm;
+    const preloads = [];
 
     while (node) {
       if (node.tagName === 'script' && this._usesAmpCacheUrl(node.attribs.src)) {
         node.attribs.src = this._replaceUrl(node.attribs.src, host);
-        if (this.esmModulesEnabled || params.experimentEsm) {
-          this._addEsm(node, node.attribs.src.endsWith('v0.js'));
+        if (esm) {
+          preloads.push(this._addEsm(node, this._shouldPreload(node.attribs.src)));
+        } else {
+          preloads.push(this._createPreload(node.attribs.src, 'script'));
         }
-        referenceNode = this._addPreload(head, referenceNode, node.attribs.src, 'script');
       } else if (
         node.tagName === 'link' &&
         node.attribs.rel === 'stylesheet' &&
         this._usesAmpCacheUrl(node.attribs.href)
       ) {
         node.attribs.href = this._replaceUrl(node.attribs.href, host);
-        referenceNode = this._addPreload(head, referenceNode, node.attribs.href, 'style');
+        preloads.push(this._createPreload(node.attribs.href, 'style'));
+      } else if (
+        node.tagName === 'link' &&
+        node.attribs.rel === 'preload' &&
+        this._usesAmpCacheUrl(node.attribs.href)
+      ) {
+        if (esm && this._shouldPreload(node.attribs.href)) {
+          // only preload mjs runtime in esm mode
+          remove(node);
+        } else {
+          node.attribs.href = this._replaceUrl(node.attribs.href, host);
+        }
       }
       node = node.nextSibling;
     }
 
+    // process preloads later to avoid accidently rewriting the URL
+    for (const preload of preloads) {
+      if (preload) {
```

Comment:
when would this be false?

Reply:
Some scripts should not have a preload. In that case rewrite URL returns `null`.